Surgery

Cardioxo opened the pull request:
### Purpose
This PR adds surgery. Related to #16

This PR still needs the following:
- [ ] Functional surgery UI after using drape.
- [ ] Basic surgery procedures working (Amputation, reattachment)
- [ ] Success rate calculation.
- [ ] Organ manipulation UI
- [ ] Advanced surgery procedures working(Organ Manipulation)

Cardioxo committed:
Surgery classes, draft

- Added most of the clases needed for surgery. Still a draft, to see what else is needed.

corp-0 commented on UnityProject/Assets/Scripts/Health/Operable.cs:
```diff
@@ -0,0 +1,78 @@
+{
+    private RegisterTile registerTile = null;
+    //The initial surgery initiation page
+    public NetTabType NetTabType = NetTabType.Surgery;
```
If the property is setteable in inspector, use [Tooltip("my tooltip")] instead of commenting the code. This will also help us know from the code alone what it does and that it is setteable in inspector.

corp-0 commented on UnityProject/Assets/Scripts/Health/Operable.cs:
```diff
@@ -0,0 +1,78 @@
+    private RegisterTile registerTile = null;
+    //The initial surgery initiation page
+    public NetTabType NetTabType = NetTabType.Surgery;
+    private HealthSystem healthSystem = null;
```
It is a style preference, but how would you feel about grouping all declarations for inspector and all privates?

corp-0 commented on UnityProject/Assets/Scripts/Health/Operable.cs:
```diff
@@ -0,0 +1,78 @@
+    
+    //Contains all the available surgery options on each body part. 
+    [SerializeField]
+    private SurgeryOptions surgeryOptions = null;
```
If the property is setteable in inspector, use [Tooltip("my tooltip")] instead of commenting the code. This will also help us know from the code alone what it does and that it is setteable in inspector.

corp-0 commented on UnityProject/Assets/Scripts/Health/Operable.cs:
```diff
@@ -0,0 +1,78 @@
+    
+    private Dictionary<BodyPartType, SurgeryProcedure> activeSurgeries = new Dictionary<BodyPartType, SurgeryProcedure>();
+
+    public Dictionary<BodyPartType, SurgeryProcedure> ActiveSurgeries
```
`public Dictionary<BodyPartType, SurgeryProcedure> ActiveSurgeries => activeSurgeries;`

corp-0 commented on UnityProject/Assets/Scripts/Health/Operable.cs:
```diff
@@ -0,0 +1,78 @@
+        EnsureInit();
+    }
+
+    private void EnsureInit()
```
if this EnsureInit is not waiting for anything, it could be inside OnEnable I guess

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SurgeryOptions.cs:
```diff
@@ -0,0 +1,15 @@
+[CreateAssetMenu(fileName = "SurgeryOptions", menuName = "ScriptableObjects/Surgery/SurgeryOptions")]
+public class SurgeryOptions : ScriptableObject
+{
+    public DictionaryBodyPartTypeToSurgeryTypeList SurgeryTypesUnderBodyPart = null;
```
Tooltip

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SurgeryOptions.cs:
```diff
@@ -0,0 +1,15 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
```
/// \<summary>

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SurgeryTools/SurgeryTool.cs:
```diff
@@ -0,0 +1,23 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+//This component allows an item to become a surgery tool.
```
/// \<summary>

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SurgeryTools/SurgeryTool.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    private SurgeryToolType toolType = SurgeryToolType.None;
+
+    public SurgeryToolType ToolType
```
``public SurgeryToolType ToolType => toolType;``

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SurgeryTools/SurgeryTool.cs:
```diff
@@ -0,0 +1,23 @@
+        get => toolType;
+    }
+    [SerializeField]
+    private int efficiency;
```
Tooltip

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SurgeryTools/SurgeryTool.cs:
```diff
@@ -0,0 +1,23 @@
+    /// <summary>
+    /// Efficiency %. The higher the efficiency, the higher the surgery speed and the higher the success rate.
+    /// </summary>
+    public float Efficiency
```
``public float Efficiency => efficiency;``

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SurgeryTools/SurgeryToolType.cs:
```diff
@@ -0,0 +1,18 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+public enum SurgeryToolType
```
/// \<summary>

corp-0 replied:
yo namespaces!

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SurgeryTypes/OrganManipulation.cs:
```diff
@@ -0,0 +1,54 @@
+{
+    void OnEnable()
+    {
+        stepAndTool = new Dictionary<int, SurgeryToolType>()
```
just thinking out loud, but since we're using integers and SurgeryToolType, we could make a SurgerySteps scriptable object with a list of SurgeryToolType and take the index of the element in the list to determine order. That way we could make new surgeries inside editor.

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SurgeryTypes/OrganManipulation.cs:
```diff
@@ -0,0 +1,54 @@
+    {
+        SurgeryToolType surgeryToolType = interaction.UsedObject.GetComponent<SurgeryTool>().ToolType;
+        //6th step is the step where organs are pulled out
+        if (CurrentStep == 6) return true;
```
if (CurrentStep == 6) 
{
    return;
}

corp-0 commented on UnityProject/Assets/Scripts/Health/Operable.cs:
```diff
@@ -0,0 +1,78 @@
+using System.Collections.Generic;
+using Mirror;
+using UnityEngine;
+using Health;
```
shouldn't this be inside Health namespace?

corp-0 commented on UnityProject/Assets/Scripts/Health/Operable.cs:
```diff
@@ -0,0 +1,78 @@
+        //Performer is self
+        if (interaction.Performer == interaction.TargetObject) return false;
+        //Target is lying down
+        if (!interaction.TargetObject.GetComponent<RegisterPlayer>().IsLayingDown) return false;
```
what if we are dissecting/operating a non-player object?

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SurgeryOptions.cs:
```diff
@@ -0,0 +1,15 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+[CreateAssetMenu(fileName = "SurgeryOptions", menuName = "ScriptableObjects/Surgery/SurgeryOptions")]
```
I still think this is health related. We could group health game objects under the same folder

Cardioxo committed:
Surgery classes, draft V2

corp-0 commented:
don't mind the comments, I forgot to submit my review last time. I will be checking this soon.

corp-0 commented on UnityProject/Assets/Scripts/Health/Operable.cs:
```diff
@@ -0,0 +1,101 @@
+    
+        private RegisterTile registerTile = null;
+        //The initial surgery initiation page
+        public NetTabType NetTabType = NetTabType.Surgery;
```
I think I already said it in the outdated comment, but I say it again. If this is for the inspector, use Tooltip instead of regular comments

corp-0 commented on UnityProject/Assets/Scripts/Health/Operable.cs:
```diff
@@ -0,0 +1,101 @@
+        public NetTabType NetTabType = NetTabType.Surgery;
+        private HealthSystem healthSystem = null;
+    
+        //Contains all the available surgery options on each body part. 
```
Tooltip

corp-0 commented on UnityProject/Assets/Scripts/Health/Operable.cs:
```diff
@@ -0,0 +1,101 @@
+            EnsureInit();
+        }
+
+        private void EnsureInit()
```
If this is not doing any operation but assigning the variables, you can do this cleanly inside OnEnable

corp-0 commented on UnityProject/Assets/Scripts/Health/Operable.cs:
```diff
@@ -0,0 +1,101 @@
+        public bool WillInteract(HandApply interaction, NetworkSide side)
+        {
+            //Within range
+            if (!DefaultWillInteract.Default(interaction, side)) return false;
```
return in a new line. If you don't want to extend it with a lot of returns, combine the conditions using ||

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SuccessEffects/Amputation.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public class Amputation : SurgeryEffect
+    {
+        public override void ApplyEffect(HandApply interaction)
```
a /// \<summary> would come in handy here

corp-0 commented on UnityProject/Assets/Scripts/Health/Surgery/SurgeryEffect.cs:
```diff
@@ -0,0 +1,10 @@
+    [CreateAssetMenu(fileName = "SurgeryEffect", menuName = "ScriptableObjects/Surgery/SurgeryEffect")]
+    public abstract class SurgeryEffect : ScriptableObject
+    {
+        public abstract void ApplyEffect(HandApply interaction);
```
/// \<summary>

Cardioxo committed:
Surgery WIP

- Added some UI png files for the surgery UI
- The Surgery net UI functionality is being worked on

- I don't know how to stash yet with git, so I'm doing this to save my current progress.
